### ISSUE TITLE
[clang-tidy] clang-tidy-diff.py reports warnings for files with incorrectly inferred compile commands

### ISSUE THREAD

**Issue by bjosv** [llvm/llvm-project#58288]
While the script `run-clang-tidy.py` uses the compilation database to know which files to run checks on,
the script `clang-tidy-diff.py` uses a unified diff to figure out which files and lines to check.

This means that `clang-tidy-diff.py` can invocate `clang-tidy` on files that are not included in the current compilation database.
It will run `clang-tidy` on headerfiles and even on source files used in different build configurations,
e.g. source files only built for certain CMake configurations or for specific platforms.

`clang-tidy` uses the `JSONCompilationDatabasePlugin` to get a known compile commands, but this is nowadays wrapped
by an `InterpolatingCompilationDatabase` class that allows it to infer compile commands for files missing in the compilation database.
This can result in a command that lacks build flags or some needed include paths.

We aim to use `clang-tidy-diff.py` as a gating step in our CI/Gerrit to detect regressions,
but we get a lot of false positives due to compilation errors, like missing include-paths.
We also get warnings when it runs on headerfiles, like unused macros or functions:
```
module/src/Functionality.h:2:9: error: macro is not used [clang-diagnostic-unused-macros]
#define MODULE_SRC_FUNCTIONALITY_H
        ^

other/include/function.h:342:24: warning: unused function 'func' [clang-diagnostic-unused-function]
static inline int func()
                  ^
```

Are we wrongly assuming that the goal of `clang-tidy-diff.py` is to detect regressions in commits as a gate,
or could this behavior be improved?

As alternatives to improve this I can think of:

- Add an option flag to `clang-tidy` to skip reporting on files that is not compilable.
  If a CompileCommand was guessed (seen via [Heuristic](https://github.com/llvm/llvm-project/blob/2e9abc0c714a97b0412de8b09b74735fafa4432b/clang/include/clang/Tooling/CompilationDatabase.h#L63-L66)?) and the diagnostics have `UncompilableErrorOccurred`
  then don't report it.

- Add an option flag to `clang-tidy-diff.py` to filter out files not available in the compilation database.

Any thoughts?
